libbeat/common/dtfmt/util.go

```diff
@@ -18,44 +18,100 @@
 package dtfmt
 
 import (
-	"math"
 	"strconv"
 )
 
 func appendUnpadded(bs []byte, i int) []byte {
 	return strconv.AppendInt(bs, int64(i), 10)
 }
 
-func appendPadded(bs []byte, i, sz int) []byte {
-	if i < 0 {
+func appendPadded(bs []byte, val, digits int) []byte {
+	if val < 0 {
 		bs = append(bs, '-')
-		i = -i
+		val = -1
 	}
 
-	if i < 10 {
-		for ; sz > 1; sz-- {
-			bs = append(bs, '0')
+	// compute number of initial padding zeroes
+	var padDigits int
+	switch {
+	case val < 10:
+		padDigits = digits - 1
+	case val < 100:
+		padDigits = digits - 2
+	case val < 1000:
+		padDigits = digits - 3
+	case val < 10000:
+		padDigits = digits - 4
+	case val < 100000:
+		padDigits = digits - 5
+	case val < 1000000:
+		padDigits = digits - 6
+	case val < 10000000:
+		padDigits = digits - 7
+	case val < 100000000:
+		padDigits = digits - 8
+	case val < 1000000000:
+		padDigits = digits - 9
+	default:
+		padDigits = digits - 1
+		for tmp := val; tmp > 10; tmp = tmp / 10 {
+			padDigits--
 		}
-		return append(bs, byte(i)+'0')
 	}
-	if i < 100 {
-		for ; sz > 2; sz-- {
-			bs = append(bs, '0')
-		}
-		return strconv.AppendInt(bs, int64(i), 10)
+	for i := 0; i < padDigits; i++ {
+		bs = append(bs, '0')
 	}
 
-	digits := 0
-	if i < 1000 {
-		digits = 3
-	} else if i < 10000 {
-		digits = 4
-	} else {
-		digits = int(math.Log10(float64(i))) + 1
+	// encode value
+	if val < 10 {
+		return append(bs, byte(val)+'0')
 	}
-	for ; sz > digits; sz-- {
-		bs = append(bs, '0')
+	return strconv.AppendInt(bs, int64(val), 10)
+}
+
+func appendFractPadded(bs []byte, val, digits, fractSz int) []byte {
```

Comment:
Could you please add more comments or a unit test to this function? I am having a hard time wrapping my head around it.

Reply:
Ok, I think I got it now. But still, a few small test cases would be nice.

Reply:
Added a comment. The function is indirectly tested via the public API tests. I'd prefer to keep blackbox testing based on the public API only.